llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp

```diff
@@ -521,10 +521,18 @@ static Value *promoteAllocaUserToVector(
       // For memset, we don't need to know the previous value because we
       // currently only allow memsets that cover the whole alloca.
       Value *Elt = MSI->getOperand(1);
-      if (DL.getTypeStoreSize(VecEltTy) > 1) {
-        Value *EltBytes =
-            Builder.CreateVectorSplat(DL.getTypeStoreSize(VecEltTy), Elt);
-        Elt = Builder.CreateBitCast(EltBytes, VecEltTy);
+      const unsigned BytesPerElt = DL.getTypeStoreSize(VecEltTy);
+      if (BytesPerElt > 1) {
+        Value *EltBytes = Builder.CreateVectorSplat(BytesPerElt, Elt);
+
+        // If the element type of the vector is a pointer, we need to first cast
+        // to an integer, then use a PtrCast.
+        if (VecEltTy->isPointerTy()) {
```

Comment:
Will fail the same way for vector of pointers

Reply:
What do you mean vector of pointers?

Reply:
this only looks at the element type of the vector type we're promoting to, which is always a primitive

Reply:
I mean [6 x <2 x ptr>]

Reply:
added the test as well, that isn't promoted either.

On a side note, PromoteAlloca needs some love still. I'm wondering if it's worth the effort to support things like nested arrays. I'd assume we'd hit the upper limit very fast with `X * Y` elements but maybe not?

Reply:
I'm actually looking into PromoteAlloca to add support for two-dimensional arrays. We have couple shaders where these will remove scratch memory accesses. I was thinking of handling two-dimensional array as multiple one-dimensional arrays, but I'm open to any suggestions. I saw @Pierre-vh you refactored and added SSAUpdater to PromoteAlloca at some point.

Reply:
It's still not trying to intelligently choose which allocas are the most profitable to promote

Reply:
@mariusz-sikora-at-amd I'm not sure, no strong opinion. I was thinking of doing it by flattening arrays (e.g. [2 x [3 x float]]) becomes [6 x float]. I think the tricky part is resolving the GEPs correctly, it might be a bigger refactoring than it looks like at first glance.

One alternative may be to have some kind of "alloca canonicalization" pass earlier that does the flattening for us to enable PromoteAlloca better.

@arsenm I haven't lost track of that but I also didn't find the time for it yet :/ 
Last time I thought about it I thought about changing the pass so it collects allocas, then sorts them by profitability (number of users + whether there are uses in loops), then just greedily promotes them one by one until it runs out of budget. Would that be good?

Reply:
Yes, that's what I was thinking for a first step

Reply:
Also, I thought SROA did try to flatten nested arrays already

Reply:
I was thinking about splitting two-dimensional array into multiple one-dimensional because:

- (this may be graphics specific) only single component / column of the array is used and by splitting we can removed unused columns / components.
- some Shaders have a very low VGPR usage, but contain "big" (more than 32 elements) arrays like [ 10 x [ 4 x float ] ]. If we split it into four one-dimensional arrays than we will be able to put everything into VGPRs.
- instruction movrel (which I'm aiming to generate) is supporting up to 32 elements.